### ISSUE TITLE
Lost focus on 'select2:open' when using 'minimumResultsForSearch' option.

### ISSUE THREAD

**Issue by nelson6e65** [select2/select2#4713]
I was having an issue similar to #3279 : I hit enter, dropdown is open, but I'm not able to navigate or search. The focus seems to go to another "place". I had to click and then it becomes navigable/search-able with keyboard.

But I think is the weird think. I another page, this works fine. 

So, I noticed that only difference was the `minimumResultsForSearch` option: 

```js
options: {
  minimumResultsForSearch: 7,
  ajax: { 
    // . . .
  }
}
```

When I remove `minimumResultsForSearch`, then works fine.

**For now, I'm disabling `minimumResultsForSearch` option**

> PD: Sorry for the incomplete/bare report, I've much time right now. But I will make some tests later and update this report with falling tests to discard or reproduce, because this option is very useful.

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [ ] The issue still exists against the latest `master` branch of Select2
- [ ] This is not a usage question (Those should be directed to the [community](https://select2.github.io/community.html))
- [ ] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Environment

Browsers

- [x] Google Chrome
- [x] Mozilla Firefox
- [ ] Internet Explorer

Operating System

- [x] Windows
- [ ] Mac OS X
- [x] Linux
- [ ] Mobile

Libraries

- jQuery version: `2.2.4`
- Select2 version: `4.0.3`

**Comment by kevin-brown** [comment]
Without an example to look at here, I'm going to guess that the focus is going to the search box, and then the search box is being hidden while it waits for results?

---

Can you expand upon the issue you are facing by adding a few more pieces of information to your bug report?

- What steps can be used to reproduce the issue?
- Does the bug happen consistently across all tested browsers?
- Can you reproduce this issue [on the examples page][examples]?
- Are you using Select2 with other plugins?

The [contributing guide][cg] contains a useful section with a few questions which can help us track down the bug and speed up the process of finding a fix for it.  The creation of [an isolated test case][test-case] would also be useful.

[cg]: https://github.com/select2/select2/blob/master/CONTRIBUTING.md#reporting-bugs-with-select2
[examples]: https://select2.github.io/examples.html
[test-case]: https://css-tricks.com/reduced-test-cases/

**Comment by shark-horse** [comment]
Reproduction of the above issue (I think) with 4.0.5 in Chrome: http://jsfiddle.net/BrfSf/404/
and with 4.0.0: http://jsfiddle.net/BrfSf/405/

Steps to reproduce the defect:
1. Create a Select2 thang with the `minimumResultsForSearch` set to fewer than the number of options.
2. Open the Select2 thang with the 'open' method of the Select2 API.

Expected: If we had a visible search box, we would focus that, but since we don't, I think we should focus the selected element in the results list, but I don't know how to use a screen reader, so what do I know? But I do know that I should be able to use up and down keys to select an item, and be able to Tab / Shift + Tab to other tabbable items on the form.

Actual: The `body` is focused, up and down do nothing, and users have to tab 1.4 million million times to get back to the selection field (estimation based on modern web site average).

Notice I'm opening the Select2 box with `$('select').select2('open')`, whereas if I click on the drop-down, or if I Tab to the drop-down and hit Space, focus is not lost and all is well.

**Comment by shark-horse** [comment]
If someone can help me understand the strategy for shifting focus, I may be able to fix this. Looking through the code, I don't see any obvious division of responsibilities regarding what components define what event listeners to pass focus onto what elements. That makes the problem seem complicated to me. :laughing:

**Comment by nelson6e65** [comment]
@shark-horse I think you are missing the `$(document).ready(function () { ... } );`

Mmm... btw, fiddles seems to work for me in Chrome. Maybe my issue was because I used the Bootstrap theme?

I don't remember. I'l check my project in witch I was using this.

**Comment by shark-horse** [comment]
Updated: http://jsfiddle.net/BrfSf/407/

Opening box with search works--I can type in the box and use the up and down arrow keys to change items:
![image](https://user-images.githubusercontent.com/30937134/32617122-c5f7779c-c539-11e7-92b5-7a30ade50c6a.png)

Opening a box without search fails. I have to click the Select box itself, or tab to it and press Space, to start using the arrow keys:
![image](https://user-images.githubusercontent.com/30937134/32617141-d15e7676-c539-11e7-9f2b-f9bde16fc6f2.png)

This is with Firefox 56.0 on Ubuntu 16.04. Same behavior (slightly different appearance) in Chrome 62.0.3202.89, same OS.

@nelson6e65 To confirm, you're not clicking on the Select2 drop-down, right? You're just calling Select2's `open` and then _immediately_ using the arrow keys?

I think I can fix this. The only place where the condition to show or not show search is handled is here:
```
var showSearch = self.showSearch(params);
if (showSearch) {
  self.$searchContainer.removeClass('select2-search--hide');
} else {
  self.$searchContainer.addClass('select2-search--hide');
}
```

The whole codebase neither queries for the `.select2-search--hide` class nor calls `showSearch` anywhere else. We should at least check whether the search box is visible (or if `showSearch()`) before focusing it.

**Comment by shark-horse** [comment]
@nelson6e65 @kevin-brown Failing test in a branch based on master submitted. I think that satisfies all the prerequisite checkboxes in the original post, but I cannot edit it.

**Comment by shark-horse** [comment]
I've been looking at and playing with select2's focus code and event handling today. It looks like focusing the Selection allows keypresses (e.g. up, down, Enter) to work as expected; that's why opening the Select2 field by UI interaction works: Whether we tab to the selection and hit Space, or click on the Select2 element, we are focusing the selection element. So it seems the solution for this defect is to focus the selection element.

It looks like the responsibility of focusing is given the component with the DOM element to focus, not e.g. the element the focus event is triggered on, i.e. everyone steals focus as they see fit. This seems problematic in the case of the dropdown search input (dropdown/search.js) specifically because it handles its own appearance / disappearance, but it has no way of knowing that the element to focus is the selection element--that could break valid extensions of Select2. Seems to me we need another way to determine who gets focus.

I could use some guidance here. Off the top of my head, I like the idea of "parent" components delegating to their "children" via API call, with `Select2` (core.js) being the topmost parent since it "knows about" all the other select field-related elements (the `select`, the selection, and the dropdown).  (Focus can be passed the other way via event triggering, perhaps?) In this example, we might say `Select2.prototype.focus` would call `focus` on either its `dropdown` or its `selection` based on the `multiple` option and whether the search element is visible... but I don't think we want `Select2` to know that much, eh? What can we do?

**Comment by shark-horse** [comment]
I pushed a fix where the dropdown search will call `.focus()` on the selection element, which it gets by JQuery-ing its `$container`. :-1: But what else can we do?

**Comment by shark-horse** [comment]
I did not figure out how to fix and reopen my previous PR, so I made a new one. Please see above.

**Comment by alexweissman** [comment]
@shark-horse can you please rebase off of 4.0.6-rc.0?  This will get folded in to the next pre-release for 4.0.6.

**Comment by shark-horse** [comment]
I rebased. Should I submit a PR to merge into the `develop` branch? I see that 4.0.6-rc.0 is not merged into `master`, so merging to `master` would make the PR include a lot of unrelated commits.

**Comment by alexweissman** [comment]
Thanks!  Also, I see you've been pretty active on the repo - do you want to drop in to my chat (https://chat.userfrosting.com/channel/select2) to talk about plans for the next release?

**Comment by shanepm** [comment]
This is a workaround I came up with for this issue (coffeescript).
```javascript
$(el).data('select2').on 'results:all', (e) =>
    context = $(el).data('select2')
    searchContainer = context.dropdown.$searchContainer
    if searchContainer.hasClass('select2-search--hide')
        context.$selection.focus()
```
